In [2]:
!pip install inspect_ai
!pip install "inspect_evals[math]@git+https://github.com/UKGovernmentBEIS/inspect_evals"

INFO: pip is looking at multiple versions of s3fs to determine which version is compatible with other requirements. This could take a while.
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.4/3.4 MB 64.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 98.2/98.2 kB 8.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 193.6/193.6 kB 15.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 135.0/135.0 kB 10.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 101.6/101.6 kB 7.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 680.1/680.1 kB 41.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 84.2/84.2 kB 7.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.7/13.7 MB 94.4 MB/s eta 0:00:00
  Attempting uninstall: fsspec
    Found existing installation: fsspec 2025.3.2
    Uninstalling fsspec-2025.3.2:
      Successfully uninstalled fsspec-2025.3.2
  Attempting uninstall: click
    Found existing installation: 

In [3]:
!pip install -e ".[dev,math]"

Obtaining file:///content
ERROR: file:///content does not appear to be a Python project: neither 'setup.py' nor 'pyproject.toml' found.


In [4]:
!inspect eval inspect_evals/arc_challenge  --model openai/gpt-4.1-nano-2025-04-14

Loading dataset allenai/ai2_arc from Hugging Face...
README.md: 9.00kB [00:00, 27.0MB/s]
ARC-Challenge/train-00000-of-00001.parqu(…): 100% 190k/190k [00:00<00:00, 1.28MB/s]
ARC-Challenge/test-00000-of-00001.parque(…): 100% 204k/204k [00:00<00:00, 1.80MB/s]
ARC-Challenge/validation-00000-of-00001.(…): 100% 55.7k/55.7k [00:00<00:00, 401kB/s]
Generating train split: 100% 1119/1119 [00:00<00:00, 25123.93 examples/s]
Generating test split: 100% 1172/1172 [00:00<00:00, 112246.52 examples/s]
Generating validation split: 100% 299/299 [00:00<00:00, 69062.00 examples/s]
Saving the dataset (1/1 shards): 100% 1172/1172 [00:00<00:00, 153520.43 examples/s]
[>1u[?2026$p[?2048$p                                                                                
  Tasks   Running Samples   Console                                             
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
                                                                                
  

dataset: allenai/ai2_arc                                                        
                                                                                
total time:                     0:01:18                                         
openai/gpt-4.1-nano-2025-04-14  128,830 tokens [I: 125,227, CW: 0, CR: 0, O:    
                              3,603, R: 0]                                      
                                                                                
accuracy: 0.89  stderr: 0.00915                                                 
                                                                                
Log: logs/2025-07-19T16-58-33+00-00_arc-challenge_QgorrhWH9XTanRng28Z2ZN.eval   

In [5]:
!inspect eval inspect_evals/arc_challenge --model openai/gpt-4o-mini-2024-07-18 --log-dir results/gpt-4o-mini --log-format json --log-level info --limit 2590

[>1u[?2026$p[?2048$p                                                                                
  Tasks   Running Samples   Console                                             
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
     

╰──────────────────────────────────────────────────────────────────────────────╯
dataset: allenai/ai2_arc                                                        
                                                                                
total time:                    0:02:03                                          
openai/gpt-4o-mini-2024-07-18  131,988 tokens [I: 125,227, CW: 0, CR: 0, O:     
                             6,761, R: 0]                                       
                                                                                
accuracy: 0.921  stderr: 0.0079                                                 
                                   

In [6]:
!inspect eval inspect_evals/arc_challenge --model openai/gpt-4.1-nano-2025-04-14 --log-dir results/gpt4.1-nano --log-format json --log-level info --limit 2590

[>1u[?2026$p[?2048$p                                                                                
  Tasks   Running Samples   Console                                             
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
     

│arc_challenge (1,172 samples): openai/gpt-4.1-nano-2025-04-14                 │
╰──────────────────────────────────────────────────────────────────────────────╯
dataset: allenai/ai2_arc                                                        
                                                                                
total time:                     0:01:53                                         
openai/gpt-4.1-nano-2025-04-14  128,743 tokens [I: 125,227, CW: 0, CR: 0, O:    
                              3,516, R: 0]                                      
                                                                                
accuracy: 0.898  stderr: 0.00883                                                
                                                                                
Log:                                                                            
results/gpt4.1-nano/2025-07-19T17-06-35+00-00_arc-challenge_VWvYbPYvm3Bk3SQFW6Eu
EU.json                                                                         
        

# 🚩 **RESULTS** 🚩


| Model         | Accuracy | Std. Error |
|---------------|----------|------------|
| **gpt-4.1-nano** | **0.898**  | **0.00883**  |
| **gpt-4o-mini**  | **0.92**   | **0.00794**  |

---

#### **gpt-4o-mini is smarter.**


In [7]:
import json

def parse_json(file_path):
    with open(file_path, 'r') as f:
        data = json.load(f)

    answers = {}
    prompts = {}
    for sample in data["samples"]:
        question_id = sample["id"]
        correct_answer = sample["target"]
        model_answer = sample["scores"]["choice"]["answer"]
        # Сам промпт
        prompt = None
        for m in sample["messages"]:
            if m["role"] == "user":
                prompt = m["content"]

        answers[question_id] = {
            "correct": correct_answer,
            "model": model_answer,
            "is_correct": (model_answer == correct_answer)
        }
        prompts[question_id] = prompt

    return answers, prompts


In [12]:
file1 = "/content/results/gpt-4o-mini/2025-07-20T05-22-57+00-00_arc-challenge_ERJpSbwha6TWtnvFaNueyU.json"
file2 = "/content/results/gpt4.1-nano/2025-07-20T05-25-00+00-00_arc-challenge_4UffAzbF9ssUXk5spN6amM.json"

In [13]:
answers_1, prompts_1 = parse_json(file1)
answers_2, prompts_2 = parse_json(file2)


only_first_correct = []
only_second_correct = []

for qid in answers_1.keys():
    first = answers_1[qid]
    second = answers_2[qid]

    if first["is_correct"] and not second["is_correct"]:
        only_first_correct.append(qid)
    elif second["is_correct"] and not first["is_correct"]:
        only_second_correct.append(qid)


print("\n--- Questions answered correctly by ONLY the FIRST model ---")
for qid in only_first_correct:
    print(f"\nID: {qid}")
    print(prompts_1[qid])


print("\n--- Questions answered correctly by ONLY the SECOND model ---")
for qid in only_second_correct:
    print(f"\nID: {qid}")
    print(prompts_1[qid])


--- Questions answered correctly by ONLY the FIRST model ---

ID: 28
Answer the following multiple choice question. The entire content of your response should be of the following format: 'ANSWER: $LETTER' (without quotes) where LETTER is one of A,B,C,D.

Students heated three objects to different temperatures during a classroom demonstration. Each object emitted light of a different color as shown below. Object 1: blue light Object 2: red light Object 3: orange light Which list presents the objects in order from highest to lowest temperature?

A) Object 1, Object 2, Object 3
B) Object 1, Object 3, Object 2
C) Object 2, Object 1, Object 3
D) Object 2, Object 3, Object 1

ID: 42
Answer the following multiple choice question. The entire content of your response should be of the following format: 'ANSWER: $LETTER' (without quotes) where LETTER is one of A,B,C,D.

Which event occurs once per day?

A) The Moon rises.
B) The new Moon occurs.
C) The Moon cycles through its phases.
D) The Moon

In [16]:
print("\n--- Questions answered correctly by ONLY the FIRST model ---")
print(" ".join(str(qid) for qid in only_first_correct))

print("\n--- Questions answered correctly by ONLY the SECOND model ---")
print(" ".join(str(qid) for qid in only_second_correct))



--- Questions answered correctly by ONLY the FIRST model ---
28 42 43 68 70 73 85 89 147 153 164 178 190 245 251 255 260 312 323 329 339 341 345 370 372 374 394 395 415 506 564 597 613 616 617 666 678 694 728 738 743 752 769 799 804 827 831 840 857 860 874 904 908 912 915 927 940 963 988 1000 1005 1065 1084 1091 1113 1117 1131 1147

--- Questions answered correctly by ONLY the SECOND model ---
6 65 88 93 168 209 234 237 250 252 268 291 311 320 336 350 408 436 437 448 452 481 512 528 670 729 748 778 796 818 819 824 865 934 994 1051 1063 1099 1137 1164 1166


1 file - 4o mini

2 file - 4.1 nano

## Comparison of the Two Models: Strengths and Differences in "Intelligence"

### **First Model Strengths:**
- **Correctly answered 61 questions vs. 31 by the second model.**
- Demonstrates **broader scientific knowledge** across multiple domains.
- Excels in **physics**: momentum, speed, power, acceleration, Newton’s laws, energy, temperature.
- Proficient in **chemistry**: atomic mass, solution concentration, phase changes.
- Strong in **experimental design**: identifying variables, understanding control groups, peer review, data interpretation.
- Confident in **biology fundamentals**: genetics, life cycles, ecology, adaptation.
- Solid performance in **earth science**: geological processes, environmental technology.
- **Type of intelligence:** A strong "encyclopedic" model — effective at recalling facts, rules, definitions, and applying textbook knowledge correctly.

---

### **Second Model Strengths:**
- **Correctly answered 31 questions, focused on biological and methodological reasoning.**
- Stronger in **scientific reasoning**: hypothesis formulation, interpreting experimental design, critique of methods.
- Performs well in **environmental science**: ecosystems, resource conservation, behaviors of animals.
- Good understanding of **disease, migration, and biological processes** (e.g., Devil Facial Tumor Disease, salmon behavior).
- Competent in **basic chemical classification, mass calculation**.
- Some isolated strengths in **physics** (air pressure, circuits), but less consistent.
- **Type of intelligence:** A "scientific thinker" — better at reasoning, causal thinking, applying the scientific method where critical analysis matters more than factual recall.

---

## **Key Difference:**

| First Model  | Second Model |
|--------------|--------------|
| More versatile and broad coverage across subjects | More focused on reasoning within specific domains |
| Superior in **quantitative reasoning**, factual recall, calculations | Superior in **scientific method, critical thinking, causal reasoning** |
| Strong in physics, chemistry, biology, earth science | Strong in biology, environmental science, experiment critique |
| Resembles a **high-performing exam taker** | Resembles a **thoughtful scientist or lab researcher** |

---

### **Summary:**
**First model is "smarter" as a comprehensive fact-based generalist.**  
**Second model is "smarter" as a scientific reasoning specialist, particularly in biology and environmental contexts.**


In [19]:
def load_answers(file_path):
    with open(file_path, 'r') as f:
        data = json.load(f)

    answers = {}
    questions = {}
    choices = {}
    for sample in data["samples"]:
        question_id = sample["id"]
        correct_answer = sample["target"]
        model_answer = sample["scores"]["choice"]["answer"]
        question_text = sample["input"]
        answer_choices = sample.get("choices", [])

        answers[question_id] = {
            "correct": correct_answer,
            "model": model_answer,
        }
        questions[question_id] = question_text
        choices[question_id] = answer_choices
    return answers, questions, choices

In [20]:
answers1, questions1, choices1 = load_answers(file1)
answers2, questions2, choices2 = load_answers(file2)

both_wrong = []

for qid in answers1:
    correct = answers1[qid]["correct"]
    answer1 = answers1[qid]["model"]
    answer2 = answers2[qid]["model"]

    if answer1 != correct and answer2 != correct:
        both_wrong.append(qid)

In [23]:
print("\nQuestions where both models got it wrong:\n")
for qid in both_wrong:
  question_text = questions1[qid]
  correct = answers1[qid]["correct"]
  answer1 = answers1[qid]["model"]
  answer2 = answers2[qid]["model"]
  answer_choices = choices1.get(qid, [])

  print(f"ID: {qid}")
  print(f"Question:\n{question_text}\n")
  if answer_choices:
    print(f"Answer options: {', '.join(answer_choices)}\n")
  print(f"Correct answer: {correct}")
  print(f"Answer of the first model: {answer1}")
  print(f"Answer of the second model: {answer2}")
  print('-' * 80)


Questions where both models got it wrong:

ID: 5
Question:
An astronaut drops a 1.0 kg object and a 5.0 kg object on the Moon. Both objects fall a total distance of 2.0 m vertically. Which of the following best describes the objects after they have fallen a distance of 1.0 m?

Answer options: They have each lost kinetic energy., They have each gained the same amount of potential energy., They have each lost the same amount of potential energy., They have each gained one-half of their maximum kinetic energy.

Correct answer: D
Answer of the first model: C
Answer of the second model: C
--------------------------------------------------------------------------------
ID: 50
Question:
Which is an example of a learned behavior?

Answer options: A tiger hunts a deer., A mouse runs from a coyote., A salmon migrates up a river., A woodpecker taps a tree for insects.

Correct answer: A
Answer of the first model: B
Answer of the second model: B
---------------------------------------------------

#Contoversial questions that should be addressed due to potential ambiguity or multiple correct answers:

1. For ID 50, the debate arises because hunting behaviors may blend learned and instinctual elements, though A (tiger hunting) is the best answer.
2. For ID 223, plate tectonics (C) could be seen as a contributing factor to mountain formation, but folded rock (A) is the most direct answer.
3. ID 466: Height and spore production are both characteristics of seedless vascular plants, but height is used for classification.  
4. ID 598: Both Volvox and Euglena are photosynthetic, but Volvox is strictly autotrophic.  
5. ID 932: Plants both photosynthesize (use carbon, release oxygen) and respire (use sugar, release carbon), but the question seems to focus on respiration.

| Name | Question | Correct Answer | Models' Answer | Why Controversial | How to Improve |
|------|----------|----------------|----------------|------------------|---------------|
| Controversial Question 50 | Example of learned behavior | A: Tiger hunts | B: Mouse flees | Hunting blends instinct and learning, causing confusion | Clarify: "Requires learning through experience"; replace instinctive options |
| Controversial Question 223 | Mountain range formation | A: Folded rock | C: Plate tectonics | Tectonics is the cause, folded rock is the result | Clarify: "Direct result of formation"; exclude tectonics |
| Controversial Question 466 | Seedless vascular plant characteristic | A: 18 cm tall | B: Produces spores | Spores and height are both traits, but height is used for classification | Clarify: "Physical trait vs. nonvascular"; replace spores |
| Controversial Question 598 | Protist requiring sunlight | A: Volvox | C: Euglena | Both Volvox and Euglena are photosynthetic, but Volvox is strictly autotrophic | Clarify: "Strictly autotrophic"; replace Euglena with non-photosynthetic protist |
| Controversial Question 932 | Plants' role in carbon cycle | B: Use sugar, release carbon | C: Use carbon, release oxygen | Photosynthesis (C) is primary, respiration (B) less expected | Clarify: "Role via respiration" or "primary role"; separate processes |

In [26]:
import pandas as pd
pd.set_option('display.max_rows', None)
df = pd.read_csv('/content/Try 2-Tаблица 1.csv', sep=';')
print(df)

      ID              Reason for Incorrect Answer by Models  \
0      5  The models likely misunderstood the physics of...   
1     50  The models likely misinterpreted "learned beha...   
2     55  The models likely associated Europa’s surface ...   
3     71  The models incorrectly chose "1" (D), possibly...   
4    140  The models chose "number of protons" (A), like...   
5    144  The models chose "Exothermic because energy is...   
6    154  The models chose "chemical energy changed to h...   
7    171  The models selected unbalanced chemical equati...   
8    191  The models chose "stem and leaf" (D), possibly...   
9    205  The models chose "surface temperature" (C), li...   
10   207  The models chose "habitat for large fish and p...   
11   223  The models chose "plate tectonics" (C), a rela...   
12   258  The models chose "greater opportunity for prim...   
13   292  The models selected unbalanced equations (D an...   
14   331  The models chose "texture" (B), possibly due 